src/gcp/cloudfunctions.ts

```diff
@@ -20,17 +33,126 @@ interface Operation {
   error?: { code: number; message: string };
 }
 
-export const API_VERSION = "v1";
+export interface HttpsTrigger {
+  // output only
+  url?: string;
+  securityLevel?: SecurityLevel;
+}
 
-export const DEFAULT_PUBLIC_POLICY = {
-  version: 3,
-  bindings: [
-    {
-      role: "roles/cloudfunctions.invoker",
-      members: ["allUsers"],
-    },
-  ],
-};
+export interface EventTrigger {
+  eventType: string;
+  resource: string;
+  service?: string;
+  failurePolicy?: FailurePolicy;
+}
+
+export interface CorsPolicy {
+  allowOrigin: string[];
+  allowMethods?: string[];
+  allowHeaders?: string[];
+  exposeHeaders?: string[];
+}
+
+export interface SecretEnvVar {
+  key: string;
+  projectId: string;
+  secret: string;
+  version: string;
+}
+
+export interface SecretVolume {
+  mountPath: string;
+  projectId: string;
+  secret: string;
+  versions: {
+    version: string;
+    path: string;
+  }[];
+}
+
+export type Runtime = "nodejs6" | "nodejs8" | "nodejs10" | "nodejs12" | "nodejs14";
+export type CloudFunctionStatus =
+  | "ACTIVE"
+  | "OFFLINE"
+  | "DEPLOY_IN_PROGRESS"
+  | "DELETE_IN_PROGRESS"
+  | "UNKNOWN";
+export type SecurityLevel = "SECURE_ALWAYS" | "SECURE_OPTIONAL";
+
+export interface FailurePolicy {
+  // oneof action
+  retry?: {};
+  // end oneof action
+}
+
+export interface CloudFunction {
+  name: string;
+  description?: string;
+
+  // oneof source_code
+  sourceArchiveUrl?: string;
+  sourceRepository?: {
+    url: string;
+    deployedUrl: string;
+  };
+  sourceUploadUrl?: string;
+  // end oneof source_code
+
+  // oneof trigger
+  httpsTrigger?: HttpsTrigger;
+  eventTrigger?: EventTrigger;
+  // end oneof trigger;
+
+  entryPoint: string;
+  runtime: Runtime;
+  // Seconds. Default = 60
+  timeout?: proto.Duration;
+
+  // Default 256
+  availableMemoryMb?: number;
+
+  // Default <projectID>@appspot.gserviceaccount.com
+  serviceAccountEmail?: string;
+
+  labels?: Record<string, string>;
+  environmentVariables?: Record<string, string>;
+  buildEnvironmentVariables?: Record<string, string>;
+
+  network?: string;
+  maxInstances?: number;
+  minInstances?: number;
+
+  corsPolicy?: CorsPolicy;
+  vpcConnector?: string;
+  vpcConnectorEgressSettings?: "PRIVATE_RANGES_ONLY" | "ALL_TRAFFIC";
+  ingressSettings?: "ALLOW_ALL" | "ALLOW_INTERNAL_ONLY" | "ALLOW_INTERNAL_AND_GCLB";
+
+  kmsKeyName?: string;
+  buildWorkerPool?: string;
+  secretEnvironmentVariables?: SecretEnvVar[];
+  secretVolumes?: SecretVolume[];
+  sourceToken?: string;
```

Comment:
Should we add a comment that `sourceToken` is input only?

Reply:
It's not though, is it? Couldn't you read this out of an existing function to pass to another?

Reply:
Strangely enough, its not returned from the API. I just double checked by adding some console logs to the functions deployment code, and sourceToken is never returned on a function by the API. For our deployment logic, we get it off of the operation metadata.